src/core/libmaven/mzUtils.cpp

```diff
@@ -1050,4 +1050,123 @@ Series:  Prentice-Hall Series in Automatic Computation
         return (strcasecmp_withNumbers(anew, bnew));
     }
 
+    double besseli0(const double x)
+    {
+        double y = 1.0;
+        double s = 1.0;
+        double n = 1.0;
+        double x2 = x * x;
+
+        while (s > y * 1.0e-9) {
+            s *= x2 / 4.0 / (n * n);
+            y += s;
+            n += 1;
+        }
+        return y;
+    }
+
+    std::vector<double> kaiser(const size_t N, const double beta)
+    {
+        std::vector<double> h(N);
+        double bb = besseli0(beta);
+        for (size_t i = 0; i < N; i++) {
+            h[i] = besseli0(beta * sqrt(4.0 * i * (N - 1 - i)) / (N - 1)) / bb;
+        }
+        return h;
+    }
+
+    double sinc(const double x)
+    {
+        // from : http://mathworld.wolfram.com/SincFunction.html
+        if (fabs(x) < 0.01)
+            return cos(M_PI * x / 2.0)
+                   * cos(M_PI * x / 4.0)
+                   * cos(M_PI * x / 8.0);
+
+        // compute sinc(x) = sin(π•x) / (π•x)
+        return sin(M_PI * x) / (M_PI * x);
+    }
+
+    std::vector<double> firDesignKaiser(const size_t n,
+                                        const double fc,
+                                        const double beta,
+                                        const double scale)
+    {
+        // validate inputs
+        if (fc < 0.0 || fc > 0.5) {
+            std::cerr << "Error: cutoff frequency "
+                      << fc
+                      << " out of range; should be within (0, 0.5)"
+                      << std::endl;
+            return {};
+        } else if (n == 0) {
+            std::cerr << "Error: filter length must be greater than zero"
+                      << std::endl;
+            return {};
+        }
+
+        // kaiser window
+        auto kw = kaiser(n, beta);
+
+        std::vector<double> filter;
+        auto t = 0.0;
+        auto h1 = 0.0;
+        auto h2 = 0.0;
+        for (size_t i = 0; i < n; i++) {
+            t = double(i) - double((n - 1) / 2.0);
+
+            // sinc prototype
+            h1 = sinc(fc * t);
+            h2 = fc * kw.at(i);
+
+            // composite
+            filter.push_back(scale * h1 * h2);
+        }
+        return filter;
+    }
+
+    std::vector<double> computeFilterCoefficients(const int interpRate,
+                                                  const int decimRate)
+    {
+        int maxRate = std::max(interpRate, decimRate);
+
+        // reasonable half-length for our sinc-like function
+        int halfLen = 10 * maxRate;
+        int numTaps = 2 * halfLen + 1;
+
+        // cutoff freq of FIR filter (rel. to Nyquist)
+        double fc = 1.0 / maxRate;
+
+        // beta for Kaiser window, 5.0 seems to work for all EICs that we tested
+        double beta = 5.0;
+
+        return firDesignKaiser(numTaps, fc, beta, interpRate);
+    }
+
+    int approximateResamplingFactor(const size_t dataSize)
+    {
+        int resamplingFactor = static_cast<int>(dataSize / 100) - 1;
```

Comment:
Considering the time range can differ a lot between experiment (20min-50min as per personal experience), just vector length of intensity might not give the best resampling factor.

A high res dataset run for 20min and a low res dataset run for 30 min might have the same vector length but if the same resampling factor is used for both, the resulting vector for the low res data might lose some features. 
I would suggest using the sample rt range as another argument for deciding the resampling factor. What do you think?

Reply:
I gave this some thought. I admit that the function responsible for calculating the resampling factor is not very robust and uses a makeshift algorithm. I agree that the sampling rate is an important factor for a simple downsampling solution, our decimation method low-pass filters a signal (to obey the Nyquist criteria) which tries to prevent aliasing (that would otherwise occur had we not filtered). Therefore, the combined method (low-pass filtering + downsampling) should be enough to allow us to disregard the time factor. This does not, however, mean that its immune to information loss or aliasing. But, to the resampling function the intensity vector is just an array values, its span is irrelevant.

If you still feel that we should consider the RT range, then we can work out some formula (but I have a feeling that it too would be ad hoc).

Reply:
Summarizing our discussion here: 
- Downsampling should not lose any information even for low-res data as you explained. It's safe to merge this in.
- Since high res data has higher data density, there is greater potential for downsampling without losing information. For this, we can use average RT difference between scans to determine the level of downsampling done for different datasets. We should test this once, but the current solution is good and can be merged.